applications/FluidDynamicsApplication/custom_conditions/navier_stokes_wall_condition.cpp

```diff
@@ -633,6 +644,133 @@ void NavierStokesWallCondition<TDim,TNumNodes>::ComputeGaussPointBehrSlipRHSCont
+{
+    KRATOS_TRY
+
+    GeometryType& rGeom = this->GetGeometry();
+
+    array_1d<double,TNumNodes> nodal_beta = ZeroVector(TNumNodes);
+	for (unsigned int nnode=0; nnode < TNumNodes; nnode++){
+        const double viscosity = rGeom[nnode].GetSolutionStepValue(DYNAMIC_VISCOSITY);
+        const double navier_slip_length = rGeom[nnode].GetValue(SLIP_LENGTH);
+        KRATOS_ERROR_IF_NOT( navier_slip_length > 0.0 ) << "Negative or zero slip length was defined" << std::endl;
+
+        nodal_beta[nnode] = viscosity / navier_slip_length;
+	}
+
+    // Retrieve the nodal consistent normal vectors, normalize, and store them for each node
+    std::vector<array_1d<double,3>> NodalNormals(TNumNodes);
+	for (unsigned int nnode=0; nnode < TNumNodes; nnode++){
+
+        NodalNormals[nnode] = rGeom[nnode].FastGetSolutionStepValue(NORMAL);
+        double sumOfSquares = 0.0;
+        for (unsigned int j = 0; j < 3; j++){
+            sumOfSquares += NodalNormals[nnode][j] * NodalNormals[nnode][j];
+        }
+        NodalNormals[nnode] = ( 1.0 / sqrt(sumOfSquares) ) *  NodalNormals[nnode];
+	}
+
+    // Computation of NodalProjectionMatrix = ( [I] - (na)(na) ) for all nodes and store it
+    std::vector< BoundedMatrix<double, TNumNodes, TNumNodes> > NodalProjectionMatrix(TNumNodes);
+    for(unsigned int node = 0; node < TNumNodes; node++){
+        FluidElementUtilities<3>::SetTangentialProjectionMatrix( NodalNormals[node], NodalProjectionMatrix[node] );
+    }
+
+    // Computation of the values for a single node
+    std::vector<array_1d<double,TNumNodes>> NodalEntriesRHS(TNumNodes);
+    for (unsigned int nnode = 0; nnode < TNumNodes; nnode++){
+
+        NodalEntriesRHS[nnode] = ZeroVector(TNumNodes);
+        const array_1d<double, TNumNodes> N = rDataStruct.N;
+        const double wGauss = rDataStruct.wGauss;
+        Vector InterpolatedTraction = ZeroVector(TNumNodes);
+        for( unsigned int comp = 0; comp < TNumNodes; comp++){
+
+            for (unsigned int i = 0; i < TNumNodes; i++){
+                // necessary because VELOCITY with 3 entries even in 2D case
+                InterpolatedTraction[i] -= N[comp] * rGeom[comp].FastGetSolutionStepValue(VELOCITY)[i];
+            }
+        }
+        // application of the nodal projection matrix
+        NodalEntriesRHS[nnode] = prod( NodalProjectionMatrix[nnode], (wGauss * N[nnode] * nodal_beta[nnode] * InterpolatedTraction) );
+    }
+
+    // putting the RHS together
+    for (unsigned int node = 0; node < TNumNodes; node++){
+        for (unsigned int entry = 0; entry < TNumNodes; entry++){
+            rRightHandSideVector( node*(TNumNodes+1) + entry ) = NodalEntriesRHS[node][entry];
+        }
+    }
+
+    KRATOS_CATCH("")
+}
+
+
+template<unsigned int TDim, unsigned int TNumNodes>
+void NavierStokesWallCondition<TDim,TNumNodes>::ComputeGaussPointNavierSlipLHSContribution( Matrix& rLeftHandSideMatrix,
```

Comment:
Same comments apply for the LHS.

Reply:
Unfortunately, here, I cannot find a shorter way. The `NodalLHSContribution` (I will  change it to `nodal_lhs_contribution`) is needed because each node has its own `NodalProjectionMatrix` (`nodal_projection_matrix`) and `nodal_beta`. Leaving out this stage and directly assigning values to `rLeftHandSideMatrix` would - in my personal opinion - make the code un-understandable. I am sorry that I cannot offer a solution at the moment.

Reply:
The vector (and loop) for the `nodal_projection_matrix` could be avoided.